core/src/impl/OpenMP/Morpheus_Multiply_Impl.hpp

```diff
@@ -0,0 +1,117 @@
+    const LinearOperator& A, const MatrixOrVector1& x, MatrixOrVector2& y,
+    CooTag, DenseVectorTag, DenseVectorTag,
+    typename std::enable_if_t<
+        Morpheus::is_execution_space_v<ExecSpace> &&
+        Morpheus::is_OpenMP_space_v<ExecSpace> &&
+        Morpheus::has_access_v<ExecSpace, LinearOperator> &&
+        Morpheus::has_access_v<ExecSpace, MatrixOrVector1> &&
+        Morpheus::has_access_v<ExecSpace, MatrixOrVector2>>* = nullptr) {
+  using IndexType = typename LinearOperator::index_type;
+
+// assumes A is sorted
+#pragma omp parallel for
+  for (IndexType n = 0; n < A.nnnz(); n++) {
+    y[A.row_indices[n]] += A.values[n] * x[A.column_indices[n]];
+  }
+}
+
+template <typename ExecSpace, typename LinearOperator, typename MatrixOrVector1,
+          typename MatrixOrVector2>
+MORPHEUS_INLINE_FUNCTION void multiply(
+    const LinearOperator& A, const MatrixOrVector1& x, MatrixOrVector2& y,
+    CsrTag, DenseVectorTag, DenseVectorTag,
+    typename std::enable_if_t<
+        Morpheus::is_execution_space_v<ExecSpace> &&
+        Morpheus::is_OpenMP_space_v<ExecSpace> &&
+        Morpheus::has_access_v<ExecSpace, LinearOperator> &&
+        Morpheus::has_access_v<ExecSpace, MatrixOrVector1> &&
+        Morpheus::has_access_v<ExecSpace, MatrixOrVector2>>* = nullptr) {
+  using IndexType = typename LinearOperator::index_type;
+  using ValueType = typename LinearOperator::value_type;
+
+#pragma omp parallel for
+  for (IndexType i = 0; i < A.nrows(); i++) {
+    ValueType sum = y[i];
+    for (IndexType jj = A.row_offsets[i]; jj < A.row_offsets[i + 1]; jj++) {
+      sum += A.values[jj] * x[A.column_indices[jj]];
+    }
+    y[i] = sum;
+  }
+}
+
+template <typename ExecSpace, typename LinearOperator, typename MatrixOrVector1,
+          typename MatrixOrVector2>
+MORPHEUS_INLINE_FUNCTION void multiply(
+    const LinearOperator& A, const MatrixOrVector1& x, MatrixOrVector2& y,
+    DiaTag, DenseVectorTag, DenseVectorTag,
+    typename std::enable_if_t<
+        Morpheus::is_execution_space_v<ExecSpace> &&
+        Morpheus::is_OpenMP_space_v<ExecSpace> &&
+        Morpheus::has_access_v<ExecSpace, LinearOperator> &&
+        Morpheus::has_access_v<ExecSpace, MatrixOrVector1> &&
+        Morpheus::has_access_v<ExecSpace, MatrixOrVector2>>* = nullptr) {
+  using IndexType       = typename LinearOperator::index_type;
+  const IndexType ndiag = A.values.ncols();
+
+  for (IndexType i = 0; i < A.nrows(); i++) y[i] = 0;
+
+#pragma omp parallel for
+  for (IndexType i = 0; i < ndiag; i++) {
+    const IndexType k = A.diagonal_offsets[i];
+
+    const IndexType i_start = std::max<IndexType>(0, -k);
+    const IndexType j_start = std::max<IndexType>(0, k);
+
+    // number of elements to process in this diagonal
+    const IndexType N = std::min(A.nrows() - i_start, A.ncols() - j_start);
+
+    for (IndexType n = 0; n < N; n++) {
+      y[i_start + n] += A.values(i_start + n, i) * x[j_start + n];
+    }
+  }
```

Comment:
y[i_start + n] will result in race condition as more than one threads might update the result at the same time. Need to find a way to work around that.